src/pip/_internal/self_outdated_check.py

```diff
@@ -225,12 +224,11 @@ def pip_self_version_check(session, options):
         if not local_version_is_older:
             return
 
-        # Advise "python -m pip" on Windows to avoid issues
-        # with overwriting pip.exe.
-        if WINDOWS:
-            pip_cmd = "python -m pip"
-        else:
-            pip_cmd = "pip"
+        # We cannot tell how the current pip is available in the current
+        # command context, so be pragmatic here and suggest the command
+        # that's always available. This doea not accomodate spaces in
+        # `sys.executable`.
+        pip_cmd = "{} -m pip".format(sys.executable)
```

Comment:
how about using pipes/shlex.quote for commands with spaces

Reply:
Neither of those work on Windows. In other situations the effect is extremely marginal (e.g. shebangs don’t work so almost no pip users would try that).

Reply:
It does work if all of the following hold:

1. generated console script
2. wheel install
3. on Unix

because we delegate to distlib which uses a [neat trick](https://bitbucket.org/pypa/distlib/src/2d145da7cb42590039fbd56a9ab764d5d4716a98/distlib/scripts.py#lines-124) to handle long and space-containing Python installation paths.

Can we make a separate issue to follow up on handling space-containing Python paths?

Reply:
Also a good reason for not just surrounding with quotes is because that would be bad advice on e.g. PowerShell, which would need something like `&'C:\path\to a\python.exe' -m pip`.

Reply:
I'm confused. We're talking about advising the user to run `{sys.executable} -m pip" and whether to quote `sys.executable`. Where do shebang lines come into the discussion?

As @chrahunt pointed out, quoting is wrong in Powershell, so blindly quoting is wrong here. Quoting on Unix may well be OK, although this will still be wrong for anyone using the cross-platform Powershell Core.

IMO, at some point we have to give up on hand-holding the user too much, and expect a certain amount of common sense. Personally, I think that suggesting `python -m pip` with a following note saying "use the path to your Python interpreter if you don't invoke Python using the `python` command" should be sufficient, But I get the impression others disagree...

Reply:
I mentioned the shebang thing as a (misguided) example that a prefix is space is problematic and might not be worthwhile to handle. Sorry for the digression.

I am not very keen on opening a new issue for the space in prefix situation is worthwhile, mostly because I don’t see there is a way to completely fix it at all, and feel it’s just as good to leave it along.

I don’t like unconditionally suggesting `python -m pip` because an uninformed user tends to simply copy-pasting whatever the tool suggests, and this can very easily break the user’s environment in a lot of cases. `sys.executable` feels like a good compromise to me; it wouldn’t always generate the right command, but at least the user will be updating the correct pip. And in the space-in-prefix situation it is almost guaranteed to not run at all (instead of running but doing the wrong thing), giving the user a chance to review the command before it takes effect.

Reply:
Agreed, `sys.executable` seems like a good compromise.